// @argent/x-shared/dist/index.js
import * as e from "./src/bigdecimal/index.js";
import { BaseError as t } from "./src/errors/baseErrors.js";
import { AccountNameGenerator as n } from "./src/accountNameGenerator/index.js";
import { AddSmartAccountResponseSchema as i, AddSmartAcountRequestSchema as s, BackendAccountSchema as c } from "./src/account/account.model.js";
import { isAccountV4 as d, isAccountV5 as p, isAccountV6 as l } from "./src/account/accountv4.js";
import { preferencesEndpointPayload as f, preferencesSchema as u } from "./src/account/preferences.model.js";
import { argentBackendNetworkIdSchema as g, argentNetworkIdSchema as x, isArgentBackendNetworkId as T, isArgentNetworkId as k } from "./src/backend/schema.js";
import { backendSessionAllowedMethodSchema as y, backendSessionMetadataSchema as D, backendSessionSchema as E, backendSessionTxFeeSchema as R } from "./src/backend/sessionKeys/schema.js";
import { CacheService as v } from "./src/cache/CacheService.js";
import { addressOrEmptyUndefinedSchema as b, addressSchema as F, addressSchemaArgentBackend as I, addressSchemaBase as B, addressSchemaLooseLength as w, addressSchemaStrictLength as _, formatFullAddress as H, formatTruncatedAddress as O, formatTruncatedSignerKey as L, formatTruncatedString as V, includesAddress as M, isAddress as W, isBase58Address as U, isEqualAddress as q, isValidAddress as G, isZeroAddress as z, normalizeAddress as K, validChecksumAddressSchema as j, validateAddressRangeSchema as X } from "./src/chains/starknet/address.js";
import { addressOrDomainInputSchema as Z, addressOrDomainSchema as Q, isEqualStarknetDomainName as J, isStarknetDomainName as $, normalizeAddressOrDomain as ee, starknetDomainNameSchema as ae } from "./src/chains/starknet/addressDomain.js";
import { addressInputCharactersAndLengthSchema as te, addressInputSchema as oe } from "./src/chains/starknet/addressInput.js";
import { argentNameSchema as me, isArgentName as ie, isEqualArgentName as se, normalizeArgentName as ce } from "./src/chains/starknet/argentName.js";
import { cairoVersionSchema as de } from "./src/chains/starknet/cairoVersionSchema.js";
import { ETH_TOKEN_ADDRESS as le, STANDARD_DEVNET_ACCOUNT_CLASS_HASH as he, STRK_TOKEN_ADDRESS as fe, TXV1_ACCOUNT_CLASS_HASH as ue, TXV1_MULTISIG_CLASS_HASH as Ae, TXV3_ACCOUNT_CLASS_HASH as ge, TXV3_MULTISIG_CLASS_HASH as xe } from "./src/chains/starknet/constants.js";
import { apiWalletResponseSchema as ke, getAddressFromArgentName as Ce } from "./src/chains/starknet/getAddressFromArgentName.js";
import { getAddressFromStarkName as De, getCallFromStarkName as Ee, getStarknetIdContractAddressFromNetworkId as Re, starkNameToCallDataDomain as Pe } from "./src/chains/starknet/getAddressFromStarkName.js";
import { getChainIdFromNetworkId as Ne } from "./src/chains/starknet/network.js";
import { StarknetAddressService as Fe } from "./src/chains/starknet/services/address/StarknetAddressService.js";
import { ARGENT_ACCOUNT_CONTRACT_CLASS_HASHES as Be, ARGENT_MULTISIG_CONTRACT_CLASS_HASHES as we, C0_PROXY_CONTRACT_CLASS_HASHES as _e } from "./src/chains/starknet/services/deploymentData/constants.js";
import { constructorCallDataSchemaC0 as Oe, constructorCallDataSchemaC1V1 as Le, implementationSchema as Ve } from "./src/chains/starknet/services/deploymentData/deploymentData.schema.js";
import { buildConstructorCalldata as We, findImplementationForAccount as Ue, getAccountContractAddress as qe, getAccountDeploymentPayload as Ge, isEqualImplementation as ze } from "./src/chains/starknet/services/deploymentData/findImplementationForAccount.js";
import { getArgentAccountClassHashes as je, getArgentAccountClassHashesWithTxV3Support as Xe, getArgentAccountWithMultiSignerClassHashes as Ye, getLastCairo0ArgentAccountClassHash as Ze, getLatestArgentAccountClassHash as Qe, getLatestArgentMultisigClassHash as Je, getLatestLedgerAccountClassHash as $e, getLedgerAccountClassHashes as ea } from "./src/chains/starknet/services/deploymentData/getAccountClassHash.js";
import { isEqualStarknetId as ra, isStarknetId as ta, normalizeStarknetId as oa, starknetIdSchema as na } from "./src/chains/starknet/starknetId.js";
import { DateService as ia } from "./src/http/DateService.js";
import { HttpError as ca } from "./src/http/HttpError.js";
import { HTTPService as da } from "./src/http/HttpService.js";
import { HTTP_ERROR_MESSAGE as la } from "./src/http/IHttpService.js";
import { SWRService as fa } from "./src/http/SWRService.js";
import { fetchData as Aa, fetcherError as ga, postData as xa } from "./src/http/fetcher.js";
import { getBackendTimeNowSeconds as ka, getBackendTimeSeconds as Ca } from "./src/http/time.js";
import { KnownDappsBackendService as Da } from "./src/knownDapps/KnownDappsBackendService.js";
import { dappContractSchema as Ra, dappLinkSchema as Pa, knownDappSchema as va, knownDappsSchema as Na } from "./src/knownDapps/schema.js";
import { getDapplandLinkForDapp as Fa, getDapplandUrlForDapp as Ia } from "./src/knownDapps/utils/getDapplandUrl.js";
import { ResolveNameService as wa } from "./src/nameResolution/ResolveNameService.js";
import { nameProviderResolutionSchema as Ha, nameResolutionAddressSchema as Oa, nameResolutionAddressesSchema as La, nameResolutionStarknetIdSchema as Va, nameResolutionStarknetIdsSchema as Ma, providersNamesSchema as Wa } from "./src/nameResolution/schema.js";
import { BackendNftService as qa } from "./src/nfts/BackendNftService.js";
import { equalNft as za, getNftPicture as Ka } from "./src/nfts/utils.js";
import { OnchainRecoveryService as Xa } from "./src/onchainRecovery/OnchainRecoveryService.js";
import { RecoveryStatusEnum as Za, StartOnchainRecoveryErrorStatusEnum as Qa, activeOnchainRecoverySchema as Ja, onchainRecoverySchema as $a, recoveryBaseSchema as er, recoveryBodySchema as ar, recoveryStatusSchema as rr, startOnchainRecoveryErrorSchema as tr, startOnchainRecoveryErrorStatusSchema as or, startOnchainRecoverySchema as nr } from "./src/onchainRecovery/schema.js";
import { isBraveBrowser as ir } from "./src/platform/isBraveBrowser.js";
import { getUAParserResult as cr, isBrowser as Sr, isChromeBrowser as dr, isEdgeBrowser as pr, isFirefoxBrowser as lr, isSafariBrowser as hr } from "./src/platform/isBrowser.js";
import { seedphraseSchema as ur } from "./src/signer/seedphrase.js";
import { MultiSigner as gr, getSignerForMultiSigner as xr } from "./src/signer/multisigner.js";
import { mergeOwnerGuardianSignatures as kr } from "./src/signer/mergeOwnerGuardianSignatures.js";
import { buildSellOptsSchema as yr, investmentMetaSchema as Dr, investmentSchema as Er, investmentsResponseSchema as Rr, lendingInvestmentSchema as Pr, liquidStakingInvestmentSchema as vr, stakerInfoSchema as Nr, stakingActionSchema as br, strkDelegatedStakingInvestmentSchema as Fr, strkStakingCalldataResponseSchema as Ir, strkStakingCalldataSchema as Br } from "./src/staking/schema.js";
import { containsValue as _r, replaceValueRecursively as Hr } from "./src/storage/utils.js";
import { mergeArrayStableWith as Lr, optionsWithDefaults as Vr } from "./src/storage/base.js";
import { getTokenBalance as Wr, getTokens as Ur, getTokensBalances as qr, useToken as Gr } from "./src/tokens/balances.js";
import { DEFAULT_TOKEN_DECIMALS as Kr } from "./src/tokens/constants.js";
import { convertTokenAmountToCurrencyValue as Xr } from "./src/tokens/convertTokenAmountToCurrencyValue.js";
import { convertTokenUnitAmountWithDecimals as Zr } from "./src/tokens/convertTokenUnitAmountWithDecimals.js";
import { ensureDecimals as Jr } from "./src/tokens/ensureDecimals.js";
import { isUnlimitedAmount as et } from "./src/tokens/isUnlimitedAmount.js";
import { prettifyCurrencyValue as rt } from "./src/tokens/prettifyCurrencyValue.js";
import { PRETTY_UNLIMITED as ot, prettifyTokenAmount as nt } from "./src/tokens/prettifyTokenAmount.js";
import { getCurrencySymbol as it } from "./src/tokens/getCurrencySymbol.js";
import { TokenServiceWeb as ct } from "./src/tokens/service/implementation.js";
import { TOKEN_ERROR_MESSAGES as dt, TokenError as pt } from "./src/tokens/service/tokenError.js";
import { apiAccountTokenBalancesSchema as ht, apiAirdropSchema as ft, apiCollateralizedDebtPositionSchema as ut, apiCollateralizedDebtPositionsSchema as At, apiConcentratedLiquidityPositionSchema as gt, apiConcentratedLiquidityPositionsSchema as xt, apiDefiDecompositionProduct as Tt, apiDefiDecompositionProducts as kt, apiDefiDecompositionSchema as Ct, apiDefiDecompositionToken as yt, apiDefiPositionSchema as Dt, apiDefiPositionsSchema as Et, apiDelegatedTokensArraySchema as Rt, apiDelegatedTokensSchema as Pt, apiPriceDataResponseSchema as vt, apiPriceDetailsSchema as Nt, apiStakingPositionSchema as bt, apiStakingPositionsSchema as Ft, apiStrkDelegatedStakingPositionsSchema as It, apiStrkDelegatedStakingSchema as Bt, apiTokenInfoResponseSchema as wt, apiTokenInfoSchema as _t, apiTokensInfoResponseSchema as Ht, apiUnclaimedRewardsSchema as Ot } from "./src/tokens/service/types/backend.model.js";
import { collateralizedDebtGroupSchema as Vt, collateralizedDebtPositionSchema as Mt, collateralizedDebtPositionsSchema as Wt, concentratedLiquidityPositionSchema as Ut, concentratedLiquidityPositionsSchema as qt, defiDappsProductsSchema as Gt, defiPositionSchema as zt, defiProductSchema as Kt, defiProductsSchema as jt, delegatedTokensArraySchema as Xt, delegatedTokensSchema as Yt, productGroupsSchema as Zt, stakingPositionSchema as Qt, stakingPositionsSchema as Jt, strkDelegatedStakingPositionArraySchema as $t, strkDelegatedStakingPositionSchema as eo } from "./src/tokens/service/types/defiPosition.model.js";
import { defiPositionTypeSchema as ro } from "./src/tokens/service/types/defiPositionType.model.js";
import { airdropSchema as oo, unclaimedRewardsSchema as no } from "./src/tokens/service/types/unclaimedRewards.model.js";
import { webBaseTokenSchema as io, webBaseTokenWithBalanceSchema as so, webTokenInfoDetailsSchema as co, webTokenPriceDetailsSchema as So, webTokenSchema as po, webTokenWithBalanceAndPriceSchema as lo, webTokenWithBalanceSchema as ho, webTokenWithPriceSchema as fo, webTokensWithBalanceAndPriceSchema as uo } from "./src/tokens/service/types/webToken.model.js";
import { isCollateralizedDebtPosition as go, isConcentratedLiquidityPosition as xo, isDelegatedTokens as To, isStaking as ko, isStrkDelegatedStaking as Co } from "./src/tokens/utils.js";
import { inputAmountSchema as Do, isValidInputAmount as Eo, parseAmount as Ro } from "./src/transactions/amount.js";
import { erc20TransferTransaction as vo, erc721SafeTransferFromTransaction as No, erc721TransferFromTransaction as bo } from "./src/transactions/buildTransactions.js";
import { getMessageFromTrpcError as Io, trpcErrorSchema as Bo } from "./src/transactions/errors.js";
import { FEE_OVERHEAD as _o, argentMaxFee as Ho, modifySnjsFeeOverhead as Oo } from "./src/transactions/estimate/argentMaxFee.js";
import { ESTIMATE_ERROR_MESSAGES as Vo, EstimateError as Mo } from "./src/transactions/estimate/error.js";
import { transactionReviewTransactionsSchema as Uo } from "./src/transactions/estimate/model.js";
import { buildDummyTx as Go, calculateCubeRootAdjustment as zo, calculateNonLinearAdjustment as Ko, calculateSqrtAdjustment as jo, castFeeEstimation as Xo, estimatedFeeToMaxFeeTotal as Yo, estimatedFeeToMaxResourceBounds as Zo, estimatedFeeToTotal as Qo, estimatedFeesToMaxFeeTotal as Jo, estimatedFeesToTotal as $o, getEstimatedFeeFromBulkSimulation as en, getEstimatedFeeFromSimulationAndRespectWatermarkFee as an, getPayloadFromTransaction as rn, getWatermarkedMaxFeeTotal as tn, isFRI as on, isWEI as nn, multiplyBigIntByFloat as mn, toMax as sn, tokensRequireTxV3Support as cn, unitToFeeTokenAddress as Sn } from "./src/transactions/estimate/utils.js";
import { EventsBySelector as pn, findTransfersAndApprovals as ln } from "./src/transactions/findTransfersAndApprovals.js";
import { getPrettyRpcError as fn } from "./src/transactions/getPrettyRcpError.js";
import { getSimulationTxVersionFromFeeToken as An, getTxVersionFromFeeToken as gn, getTxVersionFromFeeTokenForDeclareContract as xn } from "./src/transactions/getTransactionVersion.js";
import { TransactionReviewServiceWeb as kn } from "./src/transactions/service/implementation.js";
import { transactionReviewWarningSchema as yn } from "./src/transactions/service/types.js";
import { apiTransactionReviewActivityType as En } from "./src/transactions/transactionReviewTypes.js";
import { denyTxV3 as Pn, txVersionSchema as vn } from "./src/transactions/transactionVersion.js";
import { classHashSupportsTxV3 as bn, feeTokenNeedsTxV3Support as Fn } from "./src/transactions/txv3.js";
import { formatAddress as Bn } from "./src/utils/address.js";
import { ensureArray as _n } from "./src/utils/arrays.js";
import { generateAvatarImage as On, getColor as Ln, getNetworkAccountImageUrl as Vn, stripAddressZeroPadding as Mn } from "./src/utils/avatarImage.js";
import { decodeBase58 as Un, decodeBase58Array as qn, encodeBase58 as Gn, encodeBase58Array as zn } from "./src/utils/base58.js";
import { bigNumberSchema as jn } from "./src/utils/bigNumber.js";
import { daysBetween as Yn } from "./src/utils/daysBetween.js";
import { isFeatureEnabled as Qn } from "./src/utils/env.js";
import { hexSchema as $n, hexSchemaBase as em } from "./src/utils/hex.js";
import { id as rm } from "./src/utils/id.js";
import { getInitials as om } from "./src/utils/initials.js";
import { isContractDeployed as mm } from "./src/utils/isContractDeployed.js";
import { nonNullable as sm } from "./src/utils/nonNullable.js";
import { isAllowedAddressHexInputValue as Sm } from "./src/utils/number/isAllowedAddressHexInputValue.js";
import { isAllowedNumericInputValue as pm } from "./src/utils/number/isAllowedNumericInputValue.js";
import { isNumeric as hm } from "./src/utils/number/isNumeric.js";
import { prettifyCurrencyNumber as um, prettifyNumber as Am, prettifyNumberConfig as gm, prettifyTokenNumber as xm } from "./src/utils/number/prettifyNumber.js";
import { getUint256CalldataFromBN as km, parseAmountValue as Cm } from "./src/utils/parseAmount.js";
import { pluralise as Dm } from "./src/utils/pluralise.js";
import { readFileAsString as Rm } from "./src/utils/readFileAsString.js";
import { booleanToStringSchema as vm, createSchemaWithWarnings as Nm, safeParseAndWarn as bm, stringToBooleanSchema as Fm } from "./src/utils/schemas.js";
import { addOwnersCalldataSchema as Bm, addPluginCalldataSchema as wm, bigNumberishSchema as _m, callDetailsSchema as Hm, callSchema as Om, calldataSchema as Lm, changeGuardianCalldataSchema as Vm, changeThresholdCalldataSchema as Mm, escapeGuardianCalldataSchema as Wm, executeOnPluginCalldataSchema as Um, feltSchema as qm, looseCalldataSchema as Gm, multiTypeSchema as zm, rawArgsArraySchema as Km, rawArgsObjectSchema as jm, rawArgsSchema as Xm, removeOwnersCalldataSchema as Ym, removePluginCalldataSchema as Zm, replaceSignerCalldataSchema as Qm, transferCalldataSchema as Jm, uint256Schema as $m } from "./src/utils/starknet/starknet.js";
import { abiEntrySchema as ai, abiSchema as ri, builtinsSchema as ti, byteCodeSchema as oi, cairo1EventSchema as ni, cairoAssemblySchema as mi, compiledContractClassSchema as ii, compiledSierraCasmSchema as si, compiledSierraSchema as ci, compressedProgramSchema as Si, contractEntryPointFieldsSchema as di, entryPointsByTypeSchema as pi, eventAbiSchema as li, eventEntrySchema as hi, functionAbiSchema as fi, functionAbiTypeSchema as ui, legacyContractClassSchema as Ai, legacyEventSchema as gi, programSchema as xi, pythonicHintsSchema as Ti, sierraContractEntryPointFieldsSchema as ki, sierraEntryPointsByTypeSchema as Ci, sierraProgramDebugInfoSchema as yi, structAbiSchema as Di } from "./src/utils/starknet/starknetSchemas.js";
import { entryPointToHumanReadable as Ri } from "./src/utils/transactions.js";
import { urlWithQuery as vi } from "./src/utils/url/urlWithQuery.js";
import { voidify as bi } from "./src/utils/voidify.js";
import { bipsToPercent as Ii } from "./src/utils/bips.js";
import { retryUntilInitialised as wi } from "./src/utils/retryUntilInitialised.js";
export {
  Be as ARGENT_ACCOUNT_CONTRACT_CLASS_HASHES,
  we as ARGENT_MULTISIG_CONTRACT_CLASS_HASHES,
  n as AccountNameGenerator,
  i as AddSmartAccountResponseSchema,
  s as AddSmartAcountRequestSchema,
  c as BackendAccountSchema,
  qa as BackendNftService,
  t as BaseError,
  _e as C0_PROXY_CONTRACT_CLASS_HASHES,
  v as CacheService,
  Kr as DEFAULT_TOKEN_DECIMALS,
  ia as DateService,
  Vo as ESTIMATE_ERROR_MESSAGES,
  le as ETH_TOKEN_ADDRESS,
  Mo as EstimateError,
  pn as EventsBySelector,
  _o as FEE_OVERHEAD,
  da as HTTPService,
  la as HTTP_ERROR_MESSAGE,
  ca as HttpError,
  Da as KnownDappsBackendService,
  gr as MultiSigner,
  Xa as OnchainRecoveryService,
  ot as PRETTY_UNLIMITED,
  Za as RecoveryStatusEnum,
  wa as ResolveNameService,
  he as STANDARD_DEVNET_ACCOUNT_CLASS_HASH,
  fe as STRK_TOKEN_ADDRESS,
  fa as SWRService,
  Fe as StarknetAddressService,
  Qa as StartOnchainRecoveryErrorStatusEnum,
  dt as TOKEN_ERROR_MESSAGES,
  ue as TXV1_ACCOUNT_CLASS_HASH,
  Ae as TXV1_MULTISIG_CLASS_HASH,
  ge as TXV3_ACCOUNT_CLASS_HASH,
  xe as TXV3_MULTISIG_CLASS_HASH,
  pt as TokenError,
  ct as TokenServiceWeb,
  kn as TransactionReviewServiceWeb,
  ai as abiEntrySchema,
  ri as abiSchema,
  Ja as activeOnchainRecoverySchema,
  Bm as addOwnersCalldataSchema,
  wm as addPluginCalldataSchema,
  te as addressInputCharactersAndLengthSchema,
  oe as addressInputSchema,
  Z as addressOrDomainInputSchema,
  Q as addressOrDomainSchema,
  b as addressOrEmptyUndefinedSchema,
  F as addressSchema,
  I as addressSchemaArgentBackend,
  B as addressSchemaBase,
  w as addressSchemaLooseLength,
  _ as addressSchemaStrictLength,
  oo as airdropSchema,
  ht as apiAccountTokenBalancesSchema,
  ft as apiAirdropSchema,
  ut as apiCollateralizedDebtPositionSchema,
  At as apiCollateralizedDebtPositionsSchema,
  gt as apiConcentratedLiquidityPositionSchema,
  xt as apiConcentratedLiquidityPositionsSchema,
  Tt as apiDefiDecompositionProduct,
  kt as apiDefiDecompositionProducts,
  Ct as apiDefiDecompositionSchema,
  yt as apiDefiDecompositionToken,
  Dt as apiDefiPositionSchema,
  Et as apiDefiPositionsSchema,
  Rt as apiDelegatedTokensArraySchema,
  Pt as apiDelegatedTokensSchema,
  vt as apiPriceDataResponseSchema,
  Nt as apiPriceDetailsSchema,
  bt as apiStakingPositionSchema,
  Ft as apiStakingPositionsSchema,
  It as apiStrkDelegatedStakingPositionsSchema,
  Bt as apiStrkDelegatedStakingSchema,
  wt as apiTokenInfoResponseSchema,
  _t as apiTokenInfoSchema,
  Ht as apiTokensInfoResponseSchema,
  En as apiTransactionReviewActivityType,
  Ot as apiUnclaimedRewardsSchema,
  ke as apiWalletResponseSchema,
  g as argentBackendNetworkIdSchema,
  Ho as argentMaxFee,
  me as argentNameSchema,
  x as argentNetworkIdSchema,
  y as backendSessionAllowedMethodSchema,
  D as backendSessionMetadataSchema,
  E as backendSessionSchema,
  R as backendSessionTxFeeSchema,
  e as bigDecimal,
  jn as bigNumberSchema,
  _m as bigNumberishSchema,
  Ii as bipsToPercent,
  vm as booleanToStringSchema,
  We as buildConstructorCalldata,
  Go as buildDummyTx,
  yr as buildSellOptsSchema,
  ti as builtinsSchema,
  oi as byteCodeSchema,
  ni as cairo1EventSchema,
  mi as cairoAssemblySchema,
  de as cairoVersionSchema,
  zo as calculateCubeRootAdjustment,
  Ko as calculateNonLinearAdjustment,
  jo as calculateSqrtAdjustment,
  Hm as callDetailsSchema,
  Om as callSchema,
  Lm as calldataSchema,
  Xo as castFeeEstimation,
  Vm as changeGuardianCalldataSchema,
  Mm as changeThresholdCalldataSchema,
  bn as classHashSupportsTxV3,
  Vt as collateralizedDebtGroupSchema,
  Mt as collateralizedDebtPositionSchema,
  Wt as collateralizedDebtPositionsSchema,
  ii as compiledContractClassSchema,
  si as compiledSierraCasmSchema,
  ci as compiledSierraSchema,
  Si as compressedProgramSchema,
  Ut as concentratedLiquidityPositionSchema,
  qt as concentratedLiquidityPositionsSchema,
  Oe as constructorCallDataSchemaC0,
  Le as constructorCallDataSchemaC1V1,
  _r as containsValue,
  di as contractEntryPointFieldsSchema,
  Xr as convertTokenAmountToCurrencyValue,
  Zr as convertTokenUnitAmountWithDecimals,
  Nm as createSchemaWithWarnings,
  Ra as dappContractSchema,
  Pa as dappLinkSchema,
  Yn as daysBetween,
  Un as decodeBase58,
  qn as decodeBase58Array,
  Gt as defiDappsProductsSchema,
  zt as defiPositionSchema,
  ro as defiPositionTypeSchema,
  Kt as defiProductSchema,
  jt as defiProductsSchema,
  Xt as delegatedTokensArraySchema,
  Yt as delegatedTokensSchema,
  Pn as denyTxV3,
  Gn as encodeBase58,
  zn as encodeBase58Array,
  _n as ensureArray,
  Jr as ensureDecimals,
  Ri as entryPointToHumanReadable,
  pi as entryPointsByTypeSchema,
  za as equalNft,
  vo as erc20TransferTransaction,
  No as erc721SafeTransferFromTransaction,
  bo as erc721TransferFromTransaction,
  Wm as escapeGuardianCalldataSchema,
  Yo as estimatedFeeToMaxFeeTotal,
  Zo as estimatedFeeToMaxResourceBounds,
  Qo as estimatedFeeToTotal,
  Jo as estimatedFeesToMaxFeeTotal,
  $o as estimatedFeesToTotal,
  li as eventAbiSchema,
  hi as eventEntrySchema,
  Um as executeOnPluginCalldataSchema,
  Fn as feeTokenNeedsTxV3Support,
  qm as feltSchema,
  Aa as fetchData,
  ga as fetcherError,
  Ue as findImplementationForAccount,
  ln as findTransfersAndApprovals,
  Bn as formatAddress,
  H as formatFullAddress,
  O as formatTruncatedAddress,
  L as formatTruncatedSignerKey,
  V as formatTruncatedString,
  fi as functionAbiSchema,
  ui as functionAbiTypeSchema,
  On as generateAvatarImage,
  qe as getAccountContractAddress,
  Ge as getAccountDeploymentPayload,
  Ce as getAddressFromArgentName,
  De as getAddressFromStarkName,
  je as getArgentAccountClassHashes,
  Xe as getArgentAccountClassHashesWithTxV3Support,
  Ye as getArgentAccountWithMultiSignerClassHashes,
  ka as getBackendTimeNowSeconds,
  Ca as getBackendTimeSeconds,
  Ee as getCallFromStarkName,
  Ne as getChainIdFromNetworkId,
  Ln as getColor,
  it as getCurrencySymbol,
  Fa as getDapplandLinkForDapp,
  Ia as getDapplandUrlForDapp,
  en as getEstimatedFeeFromBulkSimulation,
  an as getEstimatedFeeFromSimulationAndRespectWatermarkFee,
  om as getInitials,
  Ze as getLastCairo0ArgentAccountClassHash,
  Qe as getLatestArgentAccountClassHash,
  Je as getLatestArgentMultisigClassHash,
  $e as getLatestLedgerAccountClassHash,
  ea as getLedgerAccountClassHashes,
  Io as getMessageFromTrpcError,
  Vn as getNetworkAccountImageUrl,
  Ka as getNftPicture,
  rn as getPayloadFromTransaction,
  fn as getPrettyRpcError,
  xr as getSignerForMultiSigner,
  An as getSimulationTxVersionFromFeeToken,
  Re as getStarknetIdContractAddressFromNetworkId,
  Wr as getTokenBalance,
  Ur as getTokens,
  qr as getTokensBalances,
  gn as getTxVersionFromFeeToken,
  xn as getTxVersionFromFeeTokenForDeclareContract,
  cr as getUAParserResult,
  km as getUint256CalldataFromBN,
  tn as getWatermarkedMaxFeeTotal,
  $n as hexSchema,
  em as hexSchemaBase,
  rm as id,
  Ve as implementationSchema,
  M as includesAddress,
  Do as inputAmountSchema,
  Dr as investmentMetaSchema,
  Er as investmentSchema,
  Rr as investmentsResponseSchema,
  d as isAccountV4,
  p as isAccountV5,
  l as isAccountV6,
  W as isAddress,
  Sm as isAllowedAddressHexInputValue,
  pm as isAllowedNumericInputValue,
  T as isArgentBackendNetworkId,
  ie as isArgentName,
  k as isArgentNetworkId,
  U as isBase58Address,
  ir as isBraveBrowser,
  Sr as isBrowser,
  dr as isChromeBrowser,
  go as isCollateralizedDebtPosition,
  xo as isConcentratedLiquidityPosition,
  mm as isContractDeployed,
  To as isDelegatedTokens,
  pr as isEdgeBrowser,
  q as isEqualAddress,
  se as isEqualArgentName,
  ze as isEqualImplementation,
  J as isEqualStarknetDomainName,
  ra as isEqualStarknetId,
  on as isFRI,
  Qn as isFeatureEnabled,
  lr as isFirefoxBrowser,
  hm as isNumeric,
  hr as isSafariBrowser,
  ko as isStaking,
  $ as isStarknetDomainName,
  ta as isStarknetId,
  Co as isStrkDelegatedStaking,
  et as isUnlimitedAmount,
  G as isValidAddress,
  Eo as isValidInputAmount,
  nn as isWEI,
  z as isZeroAddress,
  va as knownDappSchema,
  Na as knownDappsSchema,
  Ai as legacyContractClassSchema,
  gi as legacyEventSchema,
  Pr as lendingInvestmentSchema,
  vr as liquidStakingInvestmentSchema,
  Gm as looseCalldataSchema,
  Lr as mergeArrayStableWith,
  kr as mergeOwnerGuardianSignatures,
  Oo as modifySnjsFeeOverhead,
  zm as multiTypeSchema,
  mn as multiplyBigIntByFloat,
  Ha as nameProviderResolutionSchema,
  Oa as nameResolutionAddressSchema,
  La as nameResolutionAddressesSchema,
  Va as nameResolutionStarknetIdSchema,
  Ma as nameResolutionStarknetIdsSchema,
  sm as nonNullable,
  K as normalizeAddress,
  ee as normalizeAddressOrDomain,
  ce as normalizeArgentName,
  oa as normalizeStarknetId,
  $a as onchainRecoverySchema,
  Vr as optionsWithDefaults,
  Ro as parseAmount,
  Cm as parseAmountValue,
  Dm as pluralise,
  xa as postData,
  f as preferencesEndpointPayload,
  u as preferencesSchema,
  um as prettifyCurrencyNumber,
  rt as prettifyCurrencyValue,
  Am as prettifyNumber,
  gm as prettifyNumberConfig,
  nt as prettifyTokenAmount,
  xm as prettifyTokenNumber,
  Zt as productGroupsSchema,
  xi as programSchema,
  Wa as providersNamesSchema,
  Ti as pythonicHintsSchema,
  Km as rawArgsArraySchema,
  jm as rawArgsObjectSchema,
  Xm as rawArgsSchema,
  Rm as readFileAsString,
  er as recoveryBaseSchema,
  ar as recoveryBodySchema,
  rr as recoveryStatusSchema,
  Ym as removeOwnersCalldataSchema,
  Zm as removePluginCalldataSchema,
  Qm as replaceSignerCalldataSchema,
  Hr as replaceValueRecursively,
  wi as retryUntilInitialised,
  bm as safeParseAndWarn,
  ur as seedphraseSchema,
  ki as sierraContractEntryPointFieldsSchema,
  Ci as sierraEntryPointsByTypeSchema,
  yi as sierraProgramDebugInfoSchema,
  Nr as stakerInfoSchema,
  br as stakingActionSchema,
  Qt as stakingPositionSchema,
  Jt as stakingPositionsSchema,
  Pe as starkNameToCallDataDomain,
  ae as starknetDomainNameSchema,
  na as starknetIdSchema,
  tr as startOnchainRecoveryErrorSchema,
  or as startOnchainRecoveryErrorStatusSchema,
  nr as startOnchainRecoverySchema,
  Fm as stringToBooleanSchema,
  Mn as stripAddressZeroPadding,
  Fr as strkDelegatedStakingInvestmentSchema,
  $t as strkDelegatedStakingPositionArraySchema,
  eo as strkDelegatedStakingPositionSchema,
  Ir as strkStakingCalldataResponseSchema,
  Br as strkStakingCalldataSchema,
  Di as structAbiSchema,
  sn as toMax,
  cn as tokensRequireTxV3Support,
  Uo as transactionReviewTransactionsSchema,
  yn as transactionReviewWarningSchema,
  Jm as transferCalldataSchema,
  Bo as trpcErrorSchema,
  vn as txVersionSchema,
  $m as uint256Schema,
  no as unclaimedRewardsSchema,
  Sn as unitToFeeTokenAddress,
  vi as urlWithQuery,
  Gr as useToken,
  j as validChecksumAddressSchema,
  X as validateAddressRangeSchema,
  bi as voidify,
  io as webBaseTokenSchema,
  so as webBaseTokenWithBalanceSchema,
  co as webTokenInfoDetailsSchema,
  So as webTokenPriceDetailsSchema,
  po as webTokenSchema,
  lo as webTokenWithBalanceAndPriceSchema,
  ho as webTokenWithBalanceSchema,
  fo as webTokenWithPriceSchema,
  uo as webTokensWithBalanceAndPriceSchema
};
