// @argent/x-shared/dist/simulation.js
import { normalizeActivitySummary as i, normalizeActivitySummaryCollection as a } from "./src/features/simulation/activity/normalize.js";
import { NativeActivityStatusCancelled as m, NativeActivityStatusQueued as c, NativeActivityStatusRejected as o, NativeActivityTypeNative as s, activitiesSchema as v, activityDappSchema as n, activityDetailsSchema as S, activityResponseSchema as y, activitySchema as h, activitySummarySchema as p, activityTransferSchema as A, feeSchema as l, isActivityDetailsAction as f, isActivitySummary as u, isNftActivitySummary as x, isTokenActivitySummary as g, nativeActivityMetaSchema as d, nativeActivitySchema as N } from "./src/features/simulation/activity/schema.js";
import { createNativeActivity as T, normalizeActions as w } from "./src/features/simulation/activity/utils/createNativeActivity.js";
import { getAnyActivityShortTitle as F, getAnyActivityTitle as M } from "./src/features/simulation/activity/utils/getAnyActivityTitle.js";
import { getOverallLastModified as z } from "./src/features/simulation/activity/utils/getOverallLastModified.js";
import { hasDelegationActivity as j } from "./src/features/simulation/activity/utils/hasDelegationActivity.js";
import { isMulticallWithDeploymentActivity as O } from "./src/features/simulation/activity/utils/isMulticallWithDeploymentActivity.js";
import { isRejectOnChainActivity as k } from "./src/features/simulation/activity/utils/isRejectOnChainActivity.js";
import { isUpgradeActivity as Q } from "./src/features/simulation/activity/utils/isUpgradeActivity.js";
import { parseAccountActivities as W } from "./src/features/simulation/activity/utils/parseAccountActivities.js";
import { parseFinanceActivities as B } from "./src/features/simulation/activity/utils/parseFinanceActivities.js";
import { estimatedFeeSchema as I, estimatedFeesSchema as J } from "./src/features/simulation/fees/schema.js";
import { getErrorMessageAndLabelFromSimulation as P } from "./src/features/simulation/transactionReview/getErrorMessageAndLabelFromSimulation.js";
import { getMessageFromSimulationError as X } from "./src/features/simulation/transactionReview/getMessageFromSimulationError.js";
import { actionSchema as Z, assessmentSchema as _, enrichedSimulateAndReviewSchema as $, feeEstimationSchema as tt, isNotTransactionSimulationError as et, isTransactionSimulationError as it, propertySchema as at, reviewSchema as rt, simulateAndReviewSchema as mt, targetedDappSchema as ct, warningDetailsSchema as ot, warningSchema as st } from "./src/features/simulation/transactionReview/schema.js";
import { reasonsSchema as nt, severitySchema as St } from "./src/features/simulation/warning/schema.js";
import { getHighestSeverity as ht } from "./src/features/simulation/warning/helper.js";
export {
  m as NativeActivityStatusCancelled,
  c as NativeActivityStatusQueued,
  o as NativeActivityStatusRejected,
  s as NativeActivityTypeNative,
  Z as actionSchema,
  v as activitiesSchema,
  n as activityDappSchema,
  S as activityDetailsSchema,
  y as activityResponseSchema,
  h as activitySchema,
  p as activitySummarySchema,
  A as activityTransferSchema,
  _ as assessmentSchema,
  T as createNativeActivity,
  $ as enrichedSimulateAndReviewSchema,
  I as estimatedFeeSchema,
  J as estimatedFeesSchema,
  tt as feeEstimationSchema,
  l as feeSchema,
  F as getAnyActivityShortTitle,
  M as getAnyActivityTitle,
  P as getErrorMessageAndLabelFromSimulation,
  ht as getHighestSeverity,
  X as getMessageFromSimulationError,
  z as getOverallLastModified,
  j as hasDelegationActivity,
  f as isActivityDetailsAction,
  u as isActivitySummary,
  O as isMulticallWithDeploymentActivity,
  x as isNftActivitySummary,
  et as isNotTransactionSimulationError,
  k as isRejectOnChainActivity,
  g as isTokenActivitySummary,
  it as isTransactionSimulationError,
  Q as isUpgradeActivity,
  d as nativeActivityMetaSchema,
  N as nativeActivitySchema,
  w as normalizeActions,
  i as normalizeActivitySummary,
  a as normalizeActivitySummaryCollection,
  W as parseAccountActivities,
  B as parseFinanceActivities,
  at as propertySchema,
  nt as reasonsSchema,
  rt as reviewSchema,
  St as severitySchema,
  mt as simulateAndReviewSchema,
  ct as targetedDappSchema,
  ot as warningDetailsSchema,
  st as warningSchema
};
