// @argent/x-shared/dist/src/knownDapps/schema.js
import { z as o } from "zod";
import { addressSchemaArgentBackend as a } from "../chains/starknet/address.js";
const t = o.object({
  name: o.enum([
    "dappland",
    "discord",
    "github",
    "telegram",
    "twitter",
    "website"
  ]),
  url: o.string(),
  position: o.number()
}), n = o.object({
  address: a,
  chain: o.enum(["ethereum", "starknet"])
}), e = o.object({
  maxExpiryDays: o.number().positive(),
  allowList: o.array(o.object({
    contractAddress: a,
    methods: o.array(o.string())
  }))
}), r = o.object({
  dappId: o.string(),
  name: o.string(),
  description: o.string().optional(),
  logoUrl: o.string().optional(),
  dappUrl: o.string().optional(),
  inAppBrowserCompatible: o.boolean(),
  argentVerified: o.boolean(),
  links: o.array(t).optional(),
  contracts: o.array(n).optional(),
  categories: o.array(o.string()).optional(),
  supportedApps: o.array(o.string()).optional(),
  brandColor: o.string().optional(),
  executeFromOutsideAllowed: o.boolean().default(!1).optional(),
  sessionConfig: o.union([e, o.object({})]).nullable().optional()
}), p = o.array(r);
export {
  n as dappContractSchema,
  t as dappLinkSchema,
  r as knownDappSchema,
  p as knownDappsSchema
};
