// @argent/x-shared/dist/src/assets/tokens.json.js
const a = [
  {
    address: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
    name: "Ether",
    symbol: "ETH",
    decimals: 18,
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/eth.png"
  },
  {
    address: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
    name: "Ether",
    symbol: "ETH",
    decimals: 18,
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/eth.png"
  },
  {
    address: "0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7",
    name: "Ether",
    symbol: "ETH",
    decimals: 18,
    networkId: "localhost",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/eth.png"
  },
  {
    address: "0x04718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d",
    name: "Starknet",
    symbol: "STRK",
    decimals: "18",
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/strk.png"
  },
  {
    address: "0x04718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d",
    name: "Starknet",
    symbol: "STRK",
    decimals: "18",
    networkId: "integration",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/strk.png"
  },
  {
    address: "0x04718f5a0fc34cc1af16a1cdee98ffb20c31f5cd61d6ab07201858f4287c938d",
    name: "Starknet",
    symbol: "STRK",
    decimals: "18",
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/strk.png"
  },
  {
    address: "0x00da114221cb83fa859dbdb4c44beeaa0bb37c7537ad5ae66fe5e0efd20e6eb3",
    name: "DAI",
    symbol: "DAI",
    decimals: 18,
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/dai.png"
  },
  {
    address: "0x03e85bfbb8e2a42b7bead9e88e9a1b19dbccf661471061807292120462396ec9",
    name: "DAI",
    symbol: "DAI",
    decimals: 18,
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/dai.png"
  },
  {
    address: "0x06f2a0dfeff180133de890ad69c6ba294574c5f34a67890fd22464f348c4d03c",
    name: "DAI",
    symbol: "DAI",
    decimals: 18,
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/dai.png"
  },
  {
    address: "0x03fe2b97c1fd336e750087d68b9b867997fd64a2661ff3ca5a7c771641e8e7ac",
    name: "Wrapped BTC",
    symbol: "WBTC",
    decimals: 8,
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/wbtc.png"
  },
  {
    address: "0x12d537dc323c439dc65c976fad242d5610d27cfb5f31689a0a319b8be7f3d56",
    name: "Wrapped BTC",
    symbol: "WBTC",
    decimals: 8,
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/wbtc.png"
  },
  {
    address: "0x020d208b9e57a7f92bfa9f61135446e0961afc340378be97dbd317453c0950ae",
    name: "Wrapped BTC",
    symbol: "WBTC",
    decimals: 8,
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/wbtc.png"
  },
  {
    address: "0x053c91253bc9682c04929ca02ed00b3e423f6710d2ee7e0d5ebb06f3ecf368a8",
    name: "USD Coin",
    symbol: "USDC",
    decimals: 6,
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/usdc.png"
  },
  {
    address: "0x005a643907b9a4bc6a55e9069c4fd5fd1f5c79a22470690f75556c4736e34426",
    name: "USD Coin",
    symbol: "USDC",
    decimals: 6,
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/usdc.png"
  },
  {
    address: "0x07ab0b8855a61f480b4423c46c32fa7c553f0aac3531bbddaa282d86244f7a23",
    name: "USD Coin",
    symbol: "USDC",
    decimals: 6,
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/usdc.png"
  },
  {
    address: "0x068f5c6a61780768455de69077e07e89787839bf8166decfbf92b645209c0fb8",
    name: "Tether USD",
    symbol: "USDT",
    decimals: 6,
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/usdt.png"
  },
  {
    address: "0x386e8d061177f19b3b485c20e31137e6f6bc497cc635ccdfcab96fadf5add6a",
    name: "Tether USD",
    symbol: "USDT",
    decimals: 6,
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/usdt.png"
  },
  {
    address: "0x00585c32b625999e6e5e78645ff8df7a9001cf5cf3eb6b80ccdd16cb64bd3a34",
    name: "zkLend Token",
    symbol: "ZEND",
    decimals: 18,
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/zend.png"
  },
  {
    address: "0x06ea04cd7e40bfef07c3625e2b3472839093d966bf601568128436466f962fcf",
    name: "Baby Laika",
    symbol: "BABYLAIKA",
    decimals: 18,
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/babylaika.png"
  },
  {
    address: "0x004a8ff6a7b121c55d5e21af478265c5798f574abaa41efadf569d6939705896",
    name: "Baby Laika",
    symbol: "BABYLAIKA",
    decimals: 18,
    networkId: "sepolia-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/babylaika.png"
  },
  {
    address: "0x0782f0ddca11d9950bc3220e35ac82cf868778edb67a5e58b39838544bc4cd0f",
    name: "Starknet Voting Token",
    symbol: "vSTRK",
    decimals: 18,
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/vstrk.png"
  },
  {
    address: "0x075afe6402ad5a5c20dd25e10ec3b3986acaa647b77e4ae24b0cbc9a54a27a87",
    name: "Ekubo Protocol",
    symbol: "EKUBO",
    decimals: 18,
    networkId: "mainnet-alpha",
    image: "https://dv3jj1unlp2jl.cloudfront.net/128/color/ekubo.png"
  }
];
export {
  a as default
};
